### PR TITLE
[clang][bytecode] Start implementing __builtin_bit_cast

### PR DESCRIPTION
**Pull request by tbaederr**
This is a subset of #68288, with hopefully narrower scope. It does not support bitcasting to non-integral types yet.
Bitfields are supported, but only if they result in a full byte-sized final buffer. It does not support casting from null-pointers yet or casting from indeterminate bits.

The tests are from #68288 and partially from #74775.

The `BitcastBuffer` struct is currently always working in single bits, but I plan to (try to) optimize this for the common full-byte case.

### CODE REVIEW

File: clang/lib/AST/ByteCode/Boolean.h [hunk 1]
```diff
@@ -81,6 +81,14 @@ class Boolean final {
 
   Boolean truncate(unsigned TruncBits) const { return *this; }
 
+  static Boolean bitcastFromMemory(const std::byte *Buff, unsigned BitWidth) {
+    assert(BitWidth == 8);
```
Comment (erichkeane) [inline clang/lib/AST/ByteCode/Boolean.h:85]: Perhaps a `//FIXME: <why this is limited to 8 bits, and what we have to do to fix this>` here.

File: clang/lib/AST/ByteCode/Compiler.cpp [hunk 2]
```diff
@@ -6400,6 +6403,67 @@ bool Compiler<Emitter>::emitDestruction(const Descriptor *Desc,
...
+//  (3.4) - is_volatile_v<T> is false; and
+//  (3.5) - T has no non-static data members of reference type
+template <class Emitter>
+bool Compiler<Emitter>::emitBuiltinBitCast(const CastExpr *E) {
+  const Expr *SubExpr = E->getSubExpr();
+  QualType FromType = SubExpr->getType();
+  QualType ToType = E->getType();
+  std::optional<PrimType> ToT = classify(ToType);
+
+  assert(!DiscardResult && "Implement");
```
Comment (erichkeane) [inline clang/lib/AST/ByteCode/Compiler.cpp:6420]: would appreciate more detail here, just a little more info on the assert message itself.

File: clang/lib/AST/ByteCode/Compiler.cpp [hunk 3]
```diff
@@ -6400,6 +6403,67 @@ bool Compiler<Emitter>::emitDestruction(const Descriptor *Desc,
...
+  }
+  assert(!ToType->isReferenceType());
+
+  // Get a pointer to the value-to-cast on the stack.
+  if (!this->visit(SubExpr))
+    return false;
+
+  if (!ToT || ToT == PT_Ptr) {
+    // Conversion to an array or record type.
+    assert(false && "Implement");
```
Comment (erichkeane) [inline clang/lib/AST/ByteCode/Compiler.cpp:6443]: same here.

**Review by erichkeane — COMMENTED** [review]
Implementation seems reasonable to me, a couple of nits (and please make more detailed asserts for everything!), though I haven't paid attention to these reviews in long enough someone else should take a look.

**Comment by tbaederr** [comment]
Marking this as a draft until I have fixed things on big-endian hosts.

**Comment by tbaederr** [comment]
Victory!

 1) Iterate composite fields backwards for big-endian targets
 2) Fix `BitcatBuffer::data()` pointing to the wrong byte if the number of bytes in the buffer is not a multiple of `sizeof(uint64_t)` (which is what `llvm::BitBuffer` uses internally) and we're on a big-endian host.
 3) Validate that all of this works on little-endian and big-endian hosts, with asan and ubsan enabled.

**Comment by tbaederr** [comment]
Also, I've removed the logic for tracking uninitialized/indeterminate bits for now since it can be added later.

File: clang/lib/AST/ByteCode/Integral.h [hunk 4]
```diff
@@ -154,6 +154,18 @@ template <unsigned Bits, bool Signed> class Integral final {
...
 
+  void bitcastToMemory(std::byte *Dest) const {
+    std::memcpy(Dest, &V, sizeof(V));
+  }
+
+  static Integral bitcastFromMemory(const std::byte *Src, unsigned BitWidth) {
+    assert(BitWidth == sizeof(ReprT) * 8);
+    ReprT V;
+
+    std::memcpy(&V, Src, sizeof(ReprT));
```
Comment (AaronBallman) [inline clang/lib/AST/ByteCode/Integral.h:165]: Can we add an assert somewhere that `ReprT` will always be trivially copyable?

File: clang/test/AST/ByteCode/builtin-bit-cast.cpp [hunk 5]
```diff
@@ -0,0 +1,399 @@
...
+  union X {
+    int x;
+  };
+
+  struct G {
+    int g;
+  };
+  // expected-error@+2 {{constexpr variable 'g' must be initialized by a constant expression}}
+  // expected-note@+1 {{bit_cast from a union type is not allowed in a constant expression}}
+  constexpr G g = __builtin_bit_cast(G, X{0});
```
Comment (AaronBallman) [inline clang/test/AST/ByteCode/builtin-bit-cast.cpp:351]: Why is this test disabled? I would have assumed this worked?

**Review by AaronBallman — COMMENTED** [review]
This generally LGTM, but I'd like to see some tests for slightly more complicated situations, such as derived classes, classes with reference members, bitcasting a class with a virtual function, bitcasting a class with a virtual base, etc.

File: clang/test/AST/ByteCode/builtin-bit-cast.cpp (hunk 5 above)
Comment (tbaederr) [reply clang/test/AST/ByteCode/builtin-bit-cast.cpp:351]: It's bitcasting _to_ a non-primitive type, that's not part of this PR.

**Comment by tbaederr** [comment]
> such as derived classes,

Got me again :disappointed:

File: clang/test/AST/ByteCode/builtin-bit-cast.cpp (hunk 5 above)
Comment (AaronBallman) [reply clang/test/AST/ByteCode/builtin-bit-cast.cpp:401]: Ah! That makes sense -- can you add a test that shows bitcasting from a union to a primitive type is caught? e.g.,
```
constexpr int a = __builtin_bit_cast(int, X{0});
```

**Review by AaronBallman — APPROVED** [review]
LGTM with a bit of extra test coverage

**Comment by tbaederr** [comment]
Just tested this on a big-endian host again, no problems found.

**Pull request merged**